src/cozmo/run.py

```diff
@@ -692,3 +692,79 @@ def setup_basic_logging(general_log_level=None, protocol_log_level=None,
     if protocol_log_level is not None:
         logger_protocol.addHandler(h)
         logger_protocol.setLevel(protocol_log_level)
+
+
+class _SyncRunRobotWrapper:
+    # Wrapper to take a NON-(async/coroutine) function and
+    # wrap it so that it can be used as a run(sdk_conn: cozmo.conn.CozmoConnection)
+    # that automatically waits for a robot on the connection and then passes that
+    # to the wrapped function
+
+    def __init__(self, func_with_robot):
+        self._func_with_robot = func_with_robot
+
+    def func_with_conn(self, sdk_conn):
+        robot = sdk_conn.wait_for_robot()
+        try:
+            self._func_with_robot(robot)
+        except KeyboardInterrupt:
+            logger.info('\nExit requested by user')
+
+
+class _AsyncRunRobotWrapper:
+    # Wrapper to take an async/coroutine function and
+    # wrap it so that it can be used as a run(sdk_conn: cozmo.conn.CozmoConnection)
+    # that automatically waits for a robot on the connection and then passes that
+    # to the wrapped function
+
+    def __init__(self, func_with_robot):
+        self._func_with_robot = func_with_robot
+
+    async def func_with_conn(self, sdk_conn):
+        robot = await sdk_conn.wait_for_robot()
+        try:
+            await self._func_with_robot(robot)
+        except KeyboardInterrupt:
+            logger.info('\nExit requested by user')
+
+
+def _wrap_wait_for_robot(func_with_robot):
+    # Wrap func_with_robot (a function that takes in an already created robot)
+    # with a function that accepts a cozmo.conn.CozmoConnection
+    if asyncio.iscoroutinefunction(func_with_robot):
+        return _AsyncRunRobotWrapper(func_with_robot).func_with_conn
+    else:
+        return _SyncRunRobotWrapper(func_with_robot).func_with_conn
```

Comment:
this seems much more verbose that it needs to be... it's really just a decorator and can use the same pattern without the need for the supporting classes/objects